core/src/main/scala/com/eed3si9n/jarjarabrams/Zip.scala

```diff
@@ -1,192 +1,131 @@
-  private def directories(path: List[String]): List[String] =
-    path.foldLeft(List(""))((e, l) => (e.head + l + "/") :: e)
-
-  // map a path a/b/c to List("a/b/", "a/")
-  private def directoryPaths(path: String): List[String] =
-    directories(relativeComponents(path)).filter(_.length > 1)
-
-  // produce a sorted list of all the subdirectories of all provided files
-  private def allDirectoryPaths(files: Iterable[(Path, String)]) =
-    TreeSet[String]() ++ (files.flatMap { case (_, name) => directoryPaths(name) })
-
-  private def withZipOutput(file: Path, manifest: Option[Manifest], time: Long)(
-      f: ZipOutputStream => Unit
-  ) = {
-    Using.fileOutputStream(false)(file) { fileOut =>
-      val (zipOut, _) =
-        manifest match {
-          case Some(mf) =>
-            import Attributes.Name.MANIFEST_VERSION
-            val main = mf.getMainAttributes
-            if (!main.containsKey(MANIFEST_VERSION))
-              main.put(MANIFEST_VERSION, "1.0")
-            val os = new JarOutputStream(fileOut)
-            val e = new ZipEntry(JarFile.MANIFEST_NAME)
-            e.setTime(time)
-            os.putNextEntry(e)
-            mf.write(new BufferedOutputStream(os))
-            os.closeEntry()
-            (os, "jar")
-          case None => (new ZipOutputStream(fileOut), "zip")
-        }
-      try {
-        f(zipOut)
-      } finally {
-        zipOut.close
-      }
-    }
-  }
-
-  private def unzipStream(
-      from: InputStream,
-      toDirectory: Path,
-      preserveLastModified: Boolean = false
-  ): Set[Path] = {
-    createDirectories(toDirectory)
-    Using.zipInputStream(from)(zipInput => extract(zipInput, toDirectory, preserveLastModified))
-  }
-
-  private def extract(
-      from: ZipInputStream,
-      toDirectory: Path,
-      preserveLastModified: Boolean
-  ) = {
-    val set = new mutable.HashSet[Path]
-    @tailrec def next(): Unit = {
-      val entry = from.getNextEntry
-      if (entry == null) ()
-      else {
-        val name = entry.getName
-        val target = toDirectory.resolve(name)
-        if (entry.isDirectory) createDirectories(target)
-        else {
-          set += target
-          try {
-            Using.fileOutputStream(false)(target)(out => transfer(from, out))
-          } catch {
-            case NonFatal(e) =>
-              throw new RuntimeException(
-                "error extracting Zip entry '" + name + "' to '" + target + "'",
-                e
-              )
+  /**
+   * Treating JAR file as a set of EntryStruct, this implements a
+   * functional processing, intended for in-memory shading.
+   */
+  def transformJarFile(
+      inputJar: Path,
+      outputJar: Path,
+      resetTimestamp: Boolean
+  )(f: EntryStruct => Option[EntryStruct]): Path =
+    Using.jarFile(inputJar) { in =>
+      val tempJar = Files.createTempFile("jarjar", ".jar")
+      Using.jarOutputStream(tempJar) { out =>
```

Comment:
Why write into one jar then copy vs just write directly into the output?

Or write into a file and then on success mv it (which will often be almost free and atomic).

Reply:
Addressed in d368596fa0381745d1db9a0cf0ac80ef091f0947